### ISSUE TITLE
switch case gives different results in debug and release mode

### ISSUE THREAD

**Issue by niksen75** [dart-lang/sdk#52750]
Found a similar (Flutter) issue which has been closed, but the issue persists https://github.com/flutter/flutter/issues/125625 when switching on an enum.

### **code sample to reproduce**
enum MyEnum {
  a, 
  b, 
  c
}

String MyString(MyEnum value) {
switch (value) 
  case MyEnum.a:
    return "a";
  case MyEnum.b:
    return "b";
  case MyEnum.c:
    return "c";
  default:
    return 'default';
}

### **results**
MyString(MyEnum.a) returns 'a' both in debug and release mode
MyString(MyEnum.b) returns 'b' both in debug mode - but 'a' in release mode

### **dart --version**
Dart SDK version: 3.0.5 (stable) (Mon Jun 12 18:31:49 2023 +0000) on "macos_x64"

### **flutter doctor -v**
[✓] Flutter (Channel stable, 3.10.5, on macOS 13.4 22F66 darwin-x64, locale en-GB)
    • Flutter version 3.10.5 on channel stable at /Users/xxxx/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (8 days ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] VS Code (version 1.79.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.66.0

[✓] Connected device (3 available)
    • iPhone 13 (mobile) • 00008110-001E38E63EE2401E • ios            • iOS 16.5 20F66
    • macOS (desktop)            • macos                     • darwin-x64     • macOS 13.4 22F66 darwin-x64
    • Chrome (web)               • chrome                    • web-javascript • Google Chrome 114.0.5735.133

[✓] Network resources
    • All expected network resources are available.

**Comment by TahaTesser** [comment]
Related to https://github.com/dart-lang/sdk/issues/52214

**Comment by alexmarkov** [comment]
I'm not able to reproduce this on Dart 3.0.5.

Tried the following code (derived from the code snippet above):
```
enum MyEnum { a, b, c }

String MyString(MyEnum value) {
  switch (value) {
    case MyEnum.a:
      return "a";
    case MyEnum.b:
      return "b";
    case MyEnum.c:
      return "c";
    default:
      return 'default';
  }
}

main() {
  print(MyString(MyEnum.a));
  print(MyString(MyEnum.b));
}
```

```
dart foo.dart
a
b
```
```
dart compile exe foo.dart -o ./foo
./foo
a
b
```

@niksen75 Could you provide a complete code which reproduces the problem? Also, could you be more specific about how you build and run it?

This bug is unlikely to be related to https://github.com/dart-lang/sdk/issues/52214 / https://github.com/flutter/flutter/issues/125625, as that bug was related to the comparison of double values against int constants in the switch statements, and it didn't affect Enum values.

**Comment by niksen75** [comment]
Built in VS Code for Mac. Run it on either Simulator App (debug only) or on a physical iPhone 13 (debug and release mode). Changed the Switch Case to an If Then structure, and then it works as expected., in the current project.

However, then I also tried to run the sample code above in a clean Flutter project, and then it works as expected.

So the issue must be somehow related to the current Flutter project's implementation of the Switch Case Enum. Will investigate further, to figure out what the reason is.

But since sample code runs on Dart 3.0.5, only fair to close this ticket.
